src/v/iceberg/metadata_query.cc

```diff
@@ -0,0 +1,255 @@
+};
+
+template<>
+struct results_container<result_type::manifest> {
+    using type = chunked_vector<manifest>;
+};
+
+template<result_type ResultT>
+struct results_collector {
+    using result_t = results_container<ResultT>::type;
+    ss::stop_iteration collect(const snapshot& s) {
+        if constexpr (ResultT == result_type::snapshot) {
+            result_.push_back(s);
+            return seastar::stop_iteration::yes;
+        }
+        vassert(
+          false,
+          "Trying to collect snapshot while result type is: {}",
+          fmt::underlying(ResultT));
+    }
+    ss::stop_iteration
+    collect(const snapshot& snap, const manifest_file& m_file) {
+        if constexpr (ResultT == result_type::snapshot) {
+            result_.push_back(snap);
+            return seastar::stop_iteration::yes;
+        } else if constexpr (ResultT == result_type::manifest_file) {
+            if (!manifests_.contains(m_file.manifest_path)) {
+                manifests_.emplace(m_file.manifest_path);
+                result_.push_back(m_file.copy());
+            }
+            return seastar::stop_iteration::no;
+        }
+        vassert(
+          false,
+          "Trying to collect snapshot or manifest_file while result type is: "
+          "{}",
+          fmt::underlying(ResultT));
+    }
+    ss::stop_iteration collect(
+      const snapshot& snap,
+      const manifest_file& m_file,
+      const manifest& manifest) {
+        if constexpr (ResultT == result_type::snapshot) {
+            result_.push_back(snap);
+            return seastar::stop_iteration::yes;
+        } else if constexpr (ResultT == result_type::manifest_file) {
+            if (!manifests_.contains(m_file.manifest_path)) {
+                manifests_.emplace(m_file.manifest_path);
+                result_.push_back(m_file.copy());
+            }
+            return seastar::stop_iteration::yes;
+        } else if constexpr (ResultT == result_type::manifest) {
+            if (!manifests_.contains(m_file.manifest_path)) {
+                manifests_.emplace(m_file.manifest_path);
+                result_.push_back(manifest.copy());
+            }
+
+            return seastar::stop_iteration::no;
+        }
+        vassert(
+          false,
+          "Trying to collect snapshot, manifest_file or manifest while result "
+          "type is: {}",
+          fmt::underlying(ResultT));
+    }
+
+    results_container<ResultT>::type release() && {
+        manifests_.clear();
+        return std::move(result_);
+    }
+
+private:
+    // container used for deduplication based on path
+    chunked_hash_set<ss::sstring> manifests_;
+    results_container<ResultT>::type result_;
+};
+
+template<result_type ResultT>
+bool needs_manifest_list(const metadata_query<ResultT>& query) {
+    return query.manifest_file_matcher.has_value()
+           || query.manifest_matcher.has_value()
+           || query.r_type >= result_type::manifest_file;
```

Comment:
Is it commented anywhere the relation between the values of the enums? Maybe even worth a separate method to capture what >= is doing

Reply:
i think i will change it to simle logical or, to make it easier to understand and less controversial.